### ISSUE TITLE
OAuth2Token validation/decoding failed

### ISSUE THREAD

**Issue by Mazhar0910** [SAP/cloud-sdk#1304]
## oAuth token could not be validated/decoded , all calls are failing with 401 error

Dear Team,

I am trying to move our application to JDK 17 with Spring 3.x support, i also adopted the latest cloud sdk bom version 4.8.1-M2.
Since with latest CAP 2.x milestones and with latest cloud SDK, i could not find bean of org.springframework.security.oauth2.jwt.JwtDecoder, so i added the below configuration:

    @Bean
    **public JwtDecoder xsuaaJwtDecoder(XsuaaServiceConfiguration xsuaaServiceConfiguration) {
        return new XsuaaJwtDecoderBuilder(xsuaaServiceConfiguration).build();
    }**

    @Bean
    public XsuaaServiceConfiguration getXsuaaServiceConfiguration() {
        return new XsuaaServiceConfigurationDefault();
    }

**Now on runtime i could see all incoming calls are failing with the below error:**

"stacktrace": [
    **"java.lang.AbstractMethodError: Receiver class com.sap.cloud.security.xsuaa.token.authentication.XsuaaAudienceValidator does not define or inherit an implementation of the resolved method 'abstract org.springframework.security.oauth2.core.OAuth2TokenValidatorResult validate(org.springframework.security.oauth2.core.OAuth2Token)' of interface org.springframework.security.oauth2.core.OAuth2TokenValidator.",
    "\tat org.springframework.security.oauth2.core.DelegatingOAuth2TokenValidator.validate(DelegatingOAuth2TokenValidator.java:59)",
    "\tat org.springframework.security.oauth2.core.DelegatingOAuth2TokenValidator.validate(DelegatingOAuth2TokenValidator.java:59)",
    "\tat org.springframework.security.oauth2.jwt.NimbusJwtDecoder.validateJwt(NimbusJwtDecoder.java:185)",**
    "\tat org.springframework.security.oauth2.jwt.NimbusJwtDecoder.decode(NimbusJwtDecoder.java:138)",
    "\tat com.sap.cloud.security.xsuaa.token.authentication.XsuaaJwtDecoder.verifyWithKey(XsuaaJwtDecoder.java:173)",
    "\tat com.sap.cloud.security.xsuaa.token.authentication.XsuaaJwtDecoder.verifyToken(XsuaaJwtDecoder.java:126)",
    "\tat com.sap.cloud.security.xsuaa.token.authentication.XsuaaJwtDecoder.verifyToken(XsuaaJwtDecoder.java:111)",
    "\tat com.sap.cloud.security.xsuaa.token.authentication.XsuaaJwtDecoder.decode(XsuaaJwtDecoder.java:93)",
    "\tat org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationProvider.getJwt(JwtAuthenticationProvider.java:96)",
    "\tat org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationProvider.authenticate(JwtAuthenticationProvider.java:87)",
    "\tat org.springframework.security.authentication.ProviderManager.authenticate(ProviderManager.java:182)",
    "\tat org.springframework.security.authentication.ObservationAuthenticationManager.lambda$authenticate$1(ObservationAuthenticationManager.java:53)",
    "\tat io.micrometer.observation.Observation.observe(Observation.java:559)",
    "\tat org.springframework.security.authentication.ObservationAuthenticationManager.authenticate(ObservationAuthenticationManager.java:52)",
    "\tat org.springframework.security.oauth2.server.resource.web.authentication.BearerTokenAuthenticationFilter.doFilterInternal(BearerTokenAuthenticationFilter.java:137)",
    "\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)",

### Impact / Priority

Affected development phase

## Error Message

```
{
  "msg": "Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Filter execution threw an exception] with root cause",
  "level": "ERROR",
  "written_ts": "1678430427614916341",
  "logger": "org.apache.catalina.core.ContainerBase.[Tomcat].[localhost].[/].[dispatcherServlet]",
  "written_at": "2023-03-10T06:40:27.612Z",
  "thread": "http-nio-8080-exec-8",
  "type": "log",
  "stacktrace": [
    "java.lang.AbstractMethodError: Receiver class com.sap.cloud.security.xsuaa.token.authentication.XsuaaAudienceValidator does not define or inherit an implementation of the resolved method 'abstract org.springframework.security.oauth2.core.OAuth2TokenValidatorResult validate(org.springframework.security.oauth2.core.OAuth2Token)' of interface org.springframework.security.oauth2.core.OAuth2TokenValidator.",
    "\tat org.springframework.security.oauth2.core.DelegatingOAuth2TokenValidator.validate(DelegatingOAuth2TokenValidator.java:59)",
    "\tat org.springframework.security.oauth2.core.DelegatingOAuth2TokenValidator.validate(DelegatingOAuth2TokenValidator.java:59)",
    "\tat org.springframework.security.oauth2.jwt.NimbusJwtDecoder.validateJwt(NimbusJwtDecoder.java:185)",
    "\tat org.springframework.security.oauth2.jwt.NimbusJwtDecoder.decode(NimbusJwtDecoder.java:138)",
    "\tat com.sap.cloud.security.xsuaa.token.authentication.XsuaaJwtDecoder.verifyWithKey(XsuaaJwtDecoder.java:173)",
    "\tat com.sap.cloud.security.xsuaa.token.authentication.XsuaaJwtDecoder.verifyToken(XsuaaJwtDecoder.java:126)",
    "\tat com.sap.cloud.security.xsuaa.token.authentication.XsuaaJwtDecoder.verifyToken(XsuaaJwtDecoder.java:111)",
    "\tat com.sap.cloud.security.xsuaa.token.authentication.XsuaaJwtDecoder.decode(XsuaaJwtDecoder.java:93)",
    "\tat org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationProvider.getJwt(JwtAuthenticationProvider.java:96)",
    "\tat org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationProvider.authenticate(JwtAuthenticationProvider.java:87)",
    "\tat org.springframework.security.authentication.ProviderManager.authenticate(ProviderManager.java:182)",
    "\tat org.springframework.security.authentication.ObservationAuthenticationManager.lambda$authenticate$1(ObservationAuthenticationManager.java:53)",
    "\tat io.micrometer.observation.Observation.observe(Observation.java:559)",
    "\tat org.springframework.security.authentication.ObservationAuthenticationManager.authenticate(ObservationAuthenticationManager.java:52)",
    "\tat org.springframework.security.oauth2.server.resource.web.authentication.BearerTokenAuthenticationFilter.doFilterInternal(BearerTokenAuthenticationFilter.java:137)",
    "\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:186)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:173)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:134)",
    "\tat org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:107)",
    "\tat org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:93)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:186)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:173)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:134)",
    "\tat org.springframework.security.web.csrf.CsrfFilter.doFilterInternal(CsrfFilter.java:117)",
    "\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:186)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:173)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:134)",
    "\tat org.springframework.security.web.header.HeaderWriterFilter.doHeadersAfter(HeaderWriterFilter.java:90)",
    "\tat org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:75)",
    "\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:186)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:173)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:134)",
    "\tat org.springframework.security.web.context.SecurityContextHolderFilter.doFilter(SecurityContextHolderFilter.java:82)",
    "\tat org.springframework.security.web.context.SecurityContextHolderFilter.doFilter(SecurityContextHolderFilter.java:69)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:186)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:173)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:134)",
    "\tat org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:62)",
    "\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:186)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:173)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:134)",
    "\tat org.springframework.security.web.session.DisableEncodeUrlFilter.doFilterInternal(DisableEncodeUrlFilter.java:42)",
    "\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:186)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$AroundFilterObservation$SimpleAroundFilterObservation.lambda$wrap$0(ObservationFilterChainDecorator.java:280)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:170)",
    "\tat org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:134)",
    "\tat org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:233)",
    "\tat org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:191)",
    "\tat org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:352)",
    "\tat org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:268)",
    "\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:185)",
    "\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:158)",
    "\tat org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)",
    "\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)",
    "\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:185)",
    "\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:158)",
    "\tat org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)",
    "\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)",
    "\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:185)",
    "\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:158)",
    "\tat org.springframework.web.filter.ServerHttpObservationFilter.doFilterInternal(ServerHttpObservationFilter.java:109)",
    "\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)",
    "\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:185)",
    "\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:158)",
    "\tat org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)",
    "\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)",
    "\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:185)",
    "\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:158)",
    "\tat org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:177)",
    "\tat org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)",
    "\tat org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:542)",
    "\tat org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:119)",
    "\tat org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)",
    "\tat org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)",
    "\tat org.apache.catalina.valves.RemoteIpValve.invoke(RemoteIpValve.java:741)",
    "\tat org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:357)",
    "\tat org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:400)",
    "\tat org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)",
    "\tat org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:859)",
    "\tat org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1734)",
    "\tat org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)",
    "\tat org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)",
    "\tat org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)",
    "\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)",
    "\tat java.base/java.lang.Thread.run(Unknown Source)"
  ]
}
```

## Project Details

**https://github.tools.sap/Industry-Process-Framework/IPF/tree/JDK-17-Upgrade**

- SAP Cloud SDK Version: 4.8.1-M2
- Link to GitHub repository: [your link here](https://github.tools.sap/Industry-Process-Framework/IPF/tree/JDK-17-Upgrade)
- Project type, for example:
  - CAP Project
- Platform:
  - [Cloud Foundry

---

Thanks,
Mazhar

**Comment by Mazhar0910** [comment]
Thanks @CharlesDuboisSAP ,

I could see 4.9.0 is now available into artifactory. However the page does not display the instruction to migrate to 17 and spring boot 3

Additionally i am wondering when will XSUAA spring boot 3 compatible version shall be released. Do you have any idea ?


Thanks,
Mazhar

**Comment by CharlesDuboisSAP** [comment]
The documentation will be available when version `4.9.0`is available on Maven Central.
[The XSUAA team is planning for Q3 2022](https://github.com/SAP/cloud-security-xsuaa-integration/issues/809)

I hope this answers all your questions.
Best regards,
Charles Dubois

**Comment by Mazhar0910** [comment]
Thanks for the update :)
